engine/gamesys/src/gamesys/components/comp_particlefx.cpp

```diff
@@ -396,9 +399,43 @@ namespace dmGameSystem
                 pfx_world->m_VertexBufferData.SetSize(0);
                 pfx_world->m_RenderObjects.SetSize(0);
                 break;
-            case dmRender::RENDER_LIST_OPERATION_BATCH:
-                RenderBatch(pfx_world, params.m_Context, params.m_Buf, params.m_Begin, params.m_End);
-                break;
+            case dmRender::RENDER_LIST_OPERATION_BATCH: {
+                uint32_t *batch_begin = params.m_Begin;
+                if (RenderBatch(pfx_world, params.m_Context, params.m_Buf, &batch_begin, params.m_End)) { // will require uploading in chunks
+                    //calculate emitter sizes
+                    uint32_t vx_size = 0;
+                    for (uint32_t *i = params.m_Begin; i != params.m_End; ++i)
+                    {
+                        const dmParticle::EmitterRenderData* emitter_render_data = (dmParticle::EmitterRenderData*) params.m_Buf[*i].m_UserData;
+                        vx_size += GetEmitterVertexCount(pfx_world->m_ParticleContext, emitter_render_data->m_Instance, emitter_render_data->m_EmitterIndex);
+                    }
+
+                    //set buffer length
+                    dmGraphics::VertexAttributeInfos material_attribute_info;
+                    dmRender::HMaterial material = GetRenderMaterial(params.m_Context, (dmParticle::EmitterRenderData*)params.m_Buf[*params.m_Begin].m_UserData);
+                    dmGraphics::HVertexDeclaration vx_decl = dmRender::GetVertexDeclaration(material);
+                    FillMaterialAttributeInfos(material, vx_decl, &material_attribute_info, dmGraphics::COORDINATE_SPACE_WORLD);
+                    vx_size *= 6 * material_attribute_info.m_VertexStride;
+                    dmRender::SetBufferData(params.m_Context, pfx_world->m_VertexBuffer, vx_size, 0, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+
+                    //do the batching
+                    uint32_t vx_offset = 0;
+                    while (vx_offset < vx_size)
+                    {
+                        dmRender::SetBufferSubData(params.m_Context, pfx_world->m_VertexBuffer, vx_offset, pfx_world->m_VertexBufferData.Size(), pfx_world->m_VertexBufferData.Begin());
+                        vx_offset += pfx_world->m_VertexBufferData.Size();
+                        pfx_world->m_VertexBufferData.SetSize(0);
+                        if (!RenderBatch(pfx_world, params.m_Context, params.m_Buf, &batch_begin, params.m_End))
```

Comment:
The RenderBatch() has side effects, such as clearing the render constants and also submitting one draw call each, which is not what you wanted.

So, I suggest that you put the "looping" inside the RenderBatch() function.

Reply:
I agree, done.